### PR TITLE
Support Aux Message and Object Streaming in SP and CP

### PR DESCRIPTION
**Pull request by yanchengnv**
Fixes # .

### Description

Object streaming is built on top of Aux Messaging. However aux messaging is currently not fully implemented for SP and CP processes.

This PR adds the support of aux messaging and object streaming to SP and CP. This is necessary when, for example, the CP needs to stream log files to the SP.

The AuxRunner currently only works for CJ and SJ: the FQCNs of the message targets are hard-coded for job processes (SJ and CJ). It is modified to compute the target FQCN based on the Process Type of the current process (SP, CP, SJ, CJ).

All engine types now support aux messaging and object streaming.

Also added no_wait version of streaming in stream_runner for future use.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: nvflare/private/fed/server/fed_server.py [hunk 1]
```diff
@@ -893,7 +893,7 @@ def deploy(self, args, grpc_args=None, secure_train=False):
                     prv_key_path=grpc_args["ssl_private_key"],
                 )
 
-        self.engine.cell = self.cell
+        self.engine.initialize_comm(self.cell)
```
Comment (yhwen) [inline nvflare/private/fed/server/fed_server.py:896]: Probably better combine this initialize_comm(self.cell) with the following self._register_cellnet_cbs(). They both are doing the same for the cellnet callbacks set up.

**Review by yhwen — COMMENTED** [review]
Overall looks good. Maybe a minor change.

**Review by nvidianz — DISMISSED** [review]

**Review by YuanTingHsieh — COMMENTED** [review]
Thanks for adding the capabilities, these are great changes.
Have a few questions

File: nvflare/private/fed/server/fed_server.py (hunk 1 above)
Comment (yanchengnv) [reply nvflare/private/fed/server/fed_server.py:896]: No, this part is for SP so that the ServerEngine will get the cell.  This is also where we really need to clean up our engine(s). It's very hard to follow.

File: nvflare/private/aux_runner.py [hunk 2]
```diff
@@ -409,10 +406,25 @@ def _send_to_cell(
...
+            raise RuntimeError(f"invalid process_type {process_type}")
+
     @staticmethod
     def _convert_return_code(rc):
         rc_table = {
-            CellReturnCode.TIMEOUT: ReturnCode.COMMUNICATION_ERROR,
+            CellReturnCode.TIMEOUT: ReturnCode.TIMEOUT,
             CellReturnCode.COMM_ERROR: ReturnCode.COMMUNICATION_ERROR,
             CellReturnCode.PROCESS_EXCEPTION: ReturnCode.EXECUTION_EXCEPTION,
             CellReturnCode.ABORT_RUN: CellReturnCode.ABORT_RUN,
```
[the remark this answers is not in the document]
Comment (yanchengnv) [reply nvflare/private/aux_runner.py:430]: Don't touch this for now. The return codes are a little messy now: cell system defines a set of RCs, and FLConstant defines another set of RCs. The admin system has yet another set of RCs.  There are overlaps with different names. 

We probably should consolidate them.

File: nvflare/private/fed/server/server_engine.py [hunk 3]
```diff
@@ -586,6 +636,9 @@ def register_stream_processing(
             channel=channel, topic=topic, factory=factory, stream_done_cb=stream_done_cb, **cb_kwargs
         )
 
+    def shutdown_streamer(self):
+        self.run_manager.object_streamer.shutdown()
```
[the remark this answers is not in the document]
Comment (yanchengnv) [reply nvflare/private/fed/server/server_engine.py:640]: This is only called at the end of the whole process. But we should always be defensive so it's a good idea to check it.

File: nvflare/private/fed/server/server_runner.py [hunk 4]
```diff
@@ -218,6 +218,7 @@ def run(self):
                     self.log_info(fl_ctx, "END_RUN fired")
 
             ReliableMessage.shutdown()
+            self.engine.shutdown_streamer()
```
[the remark this answers is not in the document]
Comment (yanchengnv) [reply nvflare/private/fed/server/server_runner.py:221]: Not sure whether this particular engine has the "close" method.  Our engine is a little messy now.

File: nvflare/private/fed/server/server_engine.py [hunk 5]
```diff
@@ -381,10 +382,61 @@ def send_app_command(self, job_id: str, topic: str, cmd_data, timeout: float) ->
...
+            cb=self._handle_aux_message,
+        )
+
+    def _handle_aux_message(self, request: CellMessage) -> CellMessage:
+        assert isinstance(request, CellMessage), "request must be CellMessage but got {}".format(type(request))
+        data = request.payload
+
+        topic = request.get_header(MessageHeaderKey.TOPIC)
+        with self.new_context() as fl_ctx:
+            reply = self.run_manager.aux_runner.dispatch(topic=topic, request=data, fl_ctx=fl_ctx)
```
[the remark this answers is not in the document]
Comment (yanchengnv) [reply nvflare/private/fed/server/server_engine.py:429]: Sure, but only if our code has bug, which should be fixed.
In reality, this is impossible due to the fixed sequence of object creation (which by the way is very hard to follow).

File: nvflare/private/fed/client/client_engine.py [hunk 6]
```diff
@@ -89,6 +102,182 @@ def __init__(self, client: FederatedClient, args, rank, workers=5):
...
+            optional: whether the stream is optional
+            secure: whether to use P2P security
+
+        Returns: result from the generator's reply processing
+
+        """
+        # We are CP: can only stream to SP
+        if targets:
+            for t in targets:
+                self.logger.debug(f"ignored target: {t}")
```
[the remark this answers is not in the document]
Comment (yanchengnv) [reply nvflare/private/fed/client/client_engine.py:236]: Yes we do because this signature is required by the streaming code - it always calls the engine.stream_objects() in this way.

File: nvflare/private/fed/server/fed_server.py (hunk 1 above)
Comment (yhwen) [reply nvflare/private/fed/server/fed_server.py:896]: What I mean is these two lines could be combined. somehow HitHub comments shows the different highlight lines:

        self.engine.initialize_comm(self.cell)
        self._register_cellnet_cbs()

File: nvflare/apis/shareable.py [hunk 7]
```diff
@@ -41,9 +41,11 @@ class Shareable(dict):
     It is recommended that keys are strings. Values must be serializable.
     """
 
-    def __init__(self):
+    def __init__(self, data: dict = None):
         """Init the Shareable."""
         super().__init__()
+        if data:
+            self.update(data)
         self[ReservedHeaderKey.HEADERS] = {}
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/apis/shareable.py:49]: But not sure if this is a real issue or not, if not we can just leave it as it is.

File: nvflare/private/fed/client/client_engine.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (yanchengnv) [reply nvflare/private/fed/client/client_engine.py:239]: I added the StreamableEngine to formally specify the stream_objects() signature. and this ClientEngine implements it.

File: nvflare/apis/shareable.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (yanchengnv) [reply nvflare/apis/shareable.py:49]: That's why the special header is "reserved" - application should never set it.

**Review by YuanTingHsieh — APPROVED** [review]
LGTM, factor out streaming api to a different spec makes it clearer.
